turbopack/crates/turbopack-core/src/source_map/mod.rs

```diff
@@ -385,6 +392,100 @@ impl SourceMap {
         Ok(token.cell())
     }
 
+    /// Traces a generated line/column into an mapping token representing either
+    /// synthetic code or user-authored original code.
+    #[turbo_tasks::function]
+    pub async fn lookup_token_and_source(
+        self: Vc<Self>,
+        line: usize,
+        column: usize,
+    ) -> Result<Vc<TokenWithSource>> {
+        let mut content = None;
+
+        let token: Token = match &*self.await? {
+            SourceMap::Decoded(map) => {
+                let mut token = map
+                    .lookup_token(line as u32, column as u32)
+                    .map(Token::from)
+                    .unwrap_or_else(|| {
+                        Token::Synthetic(SyntheticToken {
+                            generated_line: line,
+                            generated_column: column,
+                            guessed_original_file: None,
+                        })
+                    });
+
+                if let Token::Synthetic(SyntheticToken {
+                    guessed_original_file,
+                    ..
+                }) = &mut token
+                {
+                    if let DecodedMap::Regular(map) = &map.map.0 {
+                        if map.get_source_count() == 1 {
+                            let source = map.sources().next().unwrap();
+                            *guessed_original_file = Some(source.to_string());
+                        }
+                    }
+                }
+
+                if let Some(map) = map.map.as_regular_source_map() {
+                    if map.get_source_count() == 1 {
```

Comment:
What's about the case when it doesn't have a single source. The `Token` points to a source index.